src/qibolab/result.py

```diff
@@ -1,100 +1,91 @@
-from dataclasses import dataclass
 from functools import cached_property
 from typing import Optional
 
 import numpy as np
 import numpy.typing as npt
 
-ExecRes = np.dtype([("i", np.float64), ("q", np.float64)])
+iq_record_dtype = np.dtype([("i", np.float64), ("q", np.float64)])
 
 
-@dataclass
 class ExecutionResults:
     """Data structure to deal with the output of :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`"""
 
-    array: npt.NDArray[ExecRes]
-    shots: Optional[npt.NDArray[np.uint32]] = None
+    def __init__(self, i: np.ndarray | list, q: np.ndarray | list, state: np.ndarray | list = None):
+        if not isinstance(i, np.ndarray):
+            i = np.array(i)
+        if not isinstance(q, np.ndarray):
+            q = np.array(q)
+        if i.shape != q.shape:
+            raise ValueError("is_ and qs_ must have the same size")
```

Comment:
Avoid this kind of checks: try to use it, and if the place where you are using it is failing, it will raise its own error.

If you want to give a more specific error, the Pythonic way would be to catch it with `try: ... except:` and reraise your more specific one (in which case consider even subclassing the error), to have a more specific type to catch somewhere else).